ai/ai21labs/v0/tasks.go

```diff
@@ -0,0 +1,387 @@
+
+func (e *execution) TaskTextGenerationChat(in *structpb.Struct) (*structpb.Struct, error) {
+	input := TaskTextGenerationChatInput{}
+	if err := base.ConvertFromStructpb(in, &input); err != nil {
+		return nil, err
+	}
+
+	messages := []ChatMessage{}
+
+	messages = append(messages, ChatMessage{
+		Role:    "system",
+		Content: input.SystemMsg,
+	})
+
+	for _, message := range input.ChatHistory {
+		contents := ""
+		for _, content := range message.Content {
+			if content.Type == "text" {
+				contents += content.Text
+			}
+		}
+		messages = append(messages, ChatMessage{
+			Role:    message.Role,
+			Content: contents,
+		})
+	}
+
+	messages = append(messages, ChatMessage{
+		Role:    "user",
+		Content: input.Prompt,
+	})
+
+	req := ChatRequest{
+		Model:       input.ModelName,
+		Messages:    messages,
+		MaxTokens:   input.MaxNewTokens,
+		Temperature: float32(input.Temperature),
+		TopP:        float32(input.TopP),
+		N:           input.N,
+	}
+
+	resp, err := e.client.Chat(req)
+	if err != nil {
+		return nil, err
+	}
+
+	outputStruct := base.TemplateTextGenerationOutput{
+		Text: resp.Choices[0].Message.Content,
+		Usage: base.GenerativeTextModelUsage{
+			InputTokens:  resp.Usage.PromptTokens,
+			OutputTokens: resp.Usage.CompletionTokens,
+		},
+	}
+
+	return base.ConvertToStructpb(outputStruct)
+}
+
+type TaskContextualAnsweringInput struct {
+	ContextualAnswersRequest
+}
+
+type TaskContextualAnsweringOutput struct {
+	Answer          string `json:"answer"`
+	AnswerInContext bool   `json:"answer-in-context"`
+}
+
+func (e *execution) TaskContextualAnswering(in *structpb.Struct) (*structpb.Struct, error) {
+	input := TaskContextualAnsweringInput{}
+	if err := base.ConvertFromStructpb(in, &input); err != nil {
+		return nil, err
+	}
+
+	resp, err := e.client.ContextualAnswers(input.ContextualAnswersRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	output := TaskContextualAnsweringOutput{
+		Answer:          resp.Answer,
+		AnswerInContext: resp.AnswerInContext,
+	}
+
+	return base.ConvertToStructpb(output)
+}
+
+type TaskTextEmbeddingsInput struct {
+	Text  string `json:"text"`
+	Style string `json:"style"`
+}
+
+type TaskTextEmbeddingsOutput struct {
+	Embedding []float32                    `json:"embedding"`
+	Usage     base.EmbeddingTextModelUsage `json:"usage"`
+}
+
+func (e *execution) TaskTextEmbeddings(in *structpb.Struct) (*structpb.Struct, error) {
+	input := TaskTextEmbeddingsInput{}
+	if err := base.ConvertFromStructpb(in, &input); err != nil {
+		return nil, err
+	}
+
+	req := EmbeddingsRequest{
+		Texts: []string{input.Text},
+		Type:  EmbeddingsType(input.Style),
+	}
+
+	resp, err := e.client.Embeddings(req)
+	if err != nil {
+		return nil, err
+	}
+
+	output := TaskTextEmbeddingsOutput{
+		Embedding: resp.Results[0].Embedding,
+		Usage: base.EmbeddingTextModelUsage{
+			Tokens: len(input.Text) / 2, // IMPORTANT: this is a rough estimate, but the embedding API does not return token counts for now (2024-07-21)
+		},
```

Comment:
https://docs.ai21.com/reference/tokenize-ref
We can use their tokenizer. But, they seem only have python one.
You can refer to [this](https://github.com/instill-ai/component/pull/270/files#diff-4195313313819939b5e147d91b90368f5d28533e0d80e0f464346dfbd47b6f44R159).

Reply:
I encountered some error:
![image](https://github.com/user-attachments/assets/c3da3c30-20ff-4596-a0ad-e6a7fc5c0415)
It seems that AI21labs themselves isn't counting the usage of embedding models at all?